### PR TITLE
asm.emu emulate also when meta is there

### PR DESCRIPTION
**Pull request by mrmacete**
- emulate silently, to avoid skipping calculating values
which could be used in subsequent instructions

also made a test for this: https://github.com/radare/radare2-regressions/pull/612

### CODE REVIEW

**Comment by radare** [comment]
what if meta says that this is data or a string? we should be more picky about that conditional i think

**Comment by radare** [comment]
oh wait. you want this thing to emulate the whole file.. we should skip the metadata parts instead of breaking or executing them

**Comment by mrmacete** [comment]
Oh i see, maybe it's my test file that's somehow malformed therefore a function is also a section. My bad. This should be solved differently. Let's clise this, but let's keep the test.

**Comment by mrmacete** [comment]
Here is my problem:

```
            ;-- section..plt:
┌ (fcn) sym._PROCEDURE_LINKAGE_TABLE_ 48
│   sym._PROCEDURE_LINKAGE_TABLE_ ();
│           0x00400820      41001c3c       lui gp, 0x41                ; [14] va=0x00400820 pa=0x00000820 sz=272 vsz=272 rwx=--r-x .plt
```
Here ^^^ the first instruction is not emulated, because it finds the meta of the  `;-- section..plt:` but actually this is also a function, so the value of `gp` on the following lines is wrong because of this

```
│           0x00400824      2454998f       lw t9, 0x5424(gp)           ; [0x422884:4]=-1 ; t9=0xffffffff -> 0xffffff00 sym.imp.__gmon_start__
│           0x00400828      24549c27       addiu gp, gp, 0x5424        ; gp=0x422884 -> 0xffffff00
│           0x0040082c      23c01c03       subu t8, t8, gp             ; t8=0xffffffffffbdd77c -> 0xffffff00
│           0x00400830      2178e003       move t7, ra                 ; t7=0x0
│           0x00400834      82c01800       srl t8, t8, 2               ; t8=0x3fef75df -> 0xffffff00
│           0x00400838      09f82003       jalr t9                     ; ra=0x400840 "A" ; $jt=0xffffffff -> 0xffffff00 sym.imp.__gmon_start__; CALL: 0x10101, 0x1, 0x0, 0x80002
│           0x0040083c      feff1827       addiu t8, t8, -2            ; t8=0x3fef75dd -> 0xffffff00
│           0x00400840      41000f3c       lui t7, 0x41                ; t7=0x410000 -> 0xffffff00
│           0x00400844      2c54f98d       lw t9, 0x542c(t7)           ; t9=0x400820 "A" sym._PROCEDURE_LINKAGE_TABLE_
│       ┌─< 0x00400848      08002003       jr t9                       ; $jt=0x400820 "A" sym._PROCEDURE_LINKAGE_TABLE_
└       │   0x0040084c      2c54f825       addiu t8, t7, 0x542c        ; t8=0x41542c -> 0x400800
```

**Comment by radare** [comment]
👍

**Comment by alvarofe** [comment]
what about emulate if the code resides in executable section?

**Comment by mrmacete** [comment]
Oh, yes. It may be faster and cleaner. Going to test that, thanks.

**Comment by radare** [comment]
Yep but what if this code is mixedup by data? This line was about to avoid emulating data inlined in code

**Comment by mrmacete** [comment]
aha, true...

**Comment by XVilka** [comment]
I think the final decision how much to emulate should be on user. So I'd suggest to add a few options regarding this and just provide reliable defaults.

**Comment by mrmacete** [comment]
Added a way to skip emulating metadata based on their type, added `asm.emuskip` config variable to let the user control that. The default is `ds` to skip just data and strings. Just squashed and rebased to avoid noise.

File: libr/core/disasm.c [hunk 1]
```diff
@@ -2993,11 +2993,11 @@ static char * resolve_fcn_name(RAnal *anal, const char * func_name) {
 	return r_anal_type_func_guess (anal, (char*)func_name);
 }
 
-static bool can_emulate_metadata(RAnal * a, ut64 at) {
+static bool can_emulate_metadata(RCore * core, ut64 at) {
```
Comment (radare) [inline libr/core/disasm.c]: handle asm.emuwrite=? to not change the value (return false) and show the help with all the possible values

**Pull request merged**